src/lua/loaders.lua

```diff
@@ -86,152 +58,268 @@ local function traverse_path(path)
+-- functions in <package.loaders> below.
+rawset(searchers, 2, searchers['path.cwd.dot'])
+rawset(searchers, 3, searchers['cpath.cwd.dot'])
+rawset(searchers, 4, searchers['path.cwd.rocks'])
+rawset(searchers, 5, searchers['cpath.cwd.rocks'])
+rawset(searchers, 6, searchers['path.package'])
+rawset(searchers, 7, searchers['cpath.package'])
+rawset(searchers, 8, searchers['croot.cwd'])
+
+-- Compose a loader function from options. The function accepts
+-- the table with the searchers and the index, so the user can
+-- override the particular searcher.
+--
+-- @param searchers table with the searcher function to be used
+--     to search a file from path template
+-- @param index in searchers table to be used in the loader to be
+--     generated
 -- @return function a loader, which first search for the file and
 --     then loads it
-local function gen_loader_func(search_fn, load_fn)
-    assert(type(search_fn) == 'function', 'search_fn must be defined')
-    assert(type(load_fn) == 'function', 'load_fn must be defined')
+local function gen_legacy_loader(searchers, index)
+    assert(type(searchers) == 'table', '<searchers> must be defined')
+    assert(type(searchers[index]) == 'function',
+           'searchers[index] must be a function')
 
     return function(name)
         if not name then
             return "empty name of module"
         end
-        local file, err = search_fn(name)
-        if not file then
-            return err
-        end
-        local loaded, err = load_fn(file, name)
-        if err == nil then
-            return loaded
-        else
-            return err
+        local loader, data = searchers[index](name)
+        -- XXX: <loader> (i.e. the first return value) contains
+        -- error message in this case. Just propagate it to the
+        -- <require> frame...
+        if not data then
+           return loader
         end
+        -- XXX: ... Otherwise, this is a valid module loader.
+        -- Load the given <data> and return the result.
+        local loaded, err = loader(data, name)
+        local message = ("error loading module '%s' from file '%s':\n\t%s")
+                        :format(name, data, err)
+        return assert(loaded, message)
     end
 end
 
-local search_lua = gen_search_func(searchroot, LUA_TEMPLATES)
-local search_lib = gen_search_func(searchroot, LIB_TEMPLATES)
-local search_rocks_lua = gen_search_func(searchroot, ROCKS_LUA_TEMPLATES, true)
-local search_rocks_lib = gen_search_func(searchroot, ROCKS_LIB_TEMPLATES, true)
-
 local search_funcs = {
-    search_lua,
-    search_lib,
-    search_rocks_lua,
-    search_rocks_lib,
-    function(name) return package.searchpath(name, package.path) end,
-    function(name) return package.searchpath(name, package.cpath) end,
+    searchers['path.cwd.dot'],
+    searchers['cpath.cwd.dot'],
+    searchers['path.cwd.rocks'],
+    searchers['cpath.cwd.rocks'],
+    searchers['path.package'],
+    searchers['cpath.package'],
 }
```

Comment:
A side note: maybe it worth to add a comment about inconsistency with the searchers/loaders machinery like one you've added for `searchers['croot.cwd']` (in fact, it is what I had to add in my patchset about override loaders).

Feel free to ignore.

Reply:
Added a comment. As for me, croot seems to be completely out of the context, since there are no Tarantool modules using this type of distribution.
```diff
diff --git a/src/lua/loaders.lua b/src/lua/loaders.lua
index c175a1b843..e9e50eea48 100644
--- a/src/lua/loaders.lua
+++ b/src/lua/loaders.lua
@@ -239,6 +239,15 @@ local function gen_legacy_loader(searchers, index)
     end
 end
 
+-- XXX: Though <package.search> looks like a handy public
+-- function, it's only the helper used for searching Tarantool
+-- C modules. Hence, despite it uses the same searchers as loaders
+-- for Lua modules do, it follows its own semantics ignoring
+-- preload, override, builtin modules, and application specific
+-- locations.
+-- FIXME: Lua paths look really odd in the list, since the
+-- function is used to load only C modules. However, these paths
+-- are left to avoid breaking change.
 local search_funcs = {
     searchers['path.cwd.dot'],
     searchers['cpath.cwd.dot'],
     ```